device/cluster.cpp

```diff
@@ -483,46 +489,49 @@ void Cluster::construct_cluster(
     noc_params = architecture_implementation->get_noc_params();
 }
 
+std::unique_ptr<Chip> Cluster::construct_chip_from_cluster(
+    chip_id_t chip_id, tt_ClusterDescriptor* cluster_desc, tt_SocDescriptor& soc_desc) {
+    if (cluster_desc->is_chip_mmio_capable(chip_id)) {
+        return std::make_unique<LocalChip>(soc_desc);
+    } else {
+        return std::make_unique<RemoteChip>(soc_desc);
+    }
+}
+
+std::unique_ptr<Chip> Cluster::construct_chip_from_cluster(chip_id_t chip_id, tt_ClusterDescriptor* cluster_desc) {
+    tt::ARCH arch = cluster_desc->get_arch(chip_id);
+    std::string soc_desc_path = tt_SocDescriptor::get_soc_descriptor_path(arch);
+    uint32_t harvesting_info = cluster_desc->get_harvesting_info().at(chip_id);
+    tt_SocDescriptor soc_desc = tt_SocDescriptor(soc_desc_path, harvesting_info);
+    return construct_chip_from_cluster(chip_id, cluster_desc, soc_desc);
+}
+
+void Cluster::add_chip(chip_id_t chip_id, std::unique_ptr<Chip> chip) {
```

Comment:
I see this is a private function, so we should use it in a proper way always, client can't misuse it, but maybe we could add a check whether chip_id is already in some of this maps. Feel free to ignore this change if you want

Reply:
Probably a good futureproof safeguard, will add it.